### ISSUE TITLE
RiichiLab ranked対局を既存GUIでlive観戦するpresentation sourceを追加する

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong-play#41]
Related:
- #25 — local AI x4 Spectator GUI
- #26 — durable-record Replay Viewer
- lisbun/lisjong-arena#223 — `lisjong-dev` RiichiLab profile / Policy mapping
- lisbun/lisjong-arena#168 — durable RiichiLab ranked game record
- lisbun/lisjong-arena#232 — bounded continuous ranked + per-hanchan durable record acquisition（COMPLETE）

# Status

**PLANNED — RiichiLab live spectator presentation source.**

`lisjong-play` には現在、次の3つのGUI presentation sourceがある。

```text
Human Play source ──┐
Replay source ───────┼─> GuiBoardRenderer / shared presentation
Local Spectator ─────┘
```

#25 のSpectatorは `lisjong-engine` 上のlocal AI x4対局をlive表示する機能であり、RiichiLab network sessionはscope外だった。

一方、Arenaには one-game ranked primitive に加え、#232で `--games N` によるbounded continuous rankedとper-hanchan durable record acquisitionが実装・live smoke済みであり、`lisjong-dev` profileから実Policyを有限回数で実戦投入できる。実運用でPolicy smokeを行う際、RiichiLab Games側には十分なリアルタイム観戦UIがなく、手元の `lisjong-play` GUIでplayer-visible stateを追いたい。

本Issueでは、**RiichiLab ranked sessionを4つ目のpresentation sourceとして既存GUIへ載せる**。

```text
Human Play source ────────┐
Replay source ─────────────┼─> GuiBoardRenderer
Local Spectator source ────┤
RiichiLab live source ──────┘
```

---

# Goal

RiichiLab rankedの1半荘をArenaのcanonical runtimeで実行しながら、lisjong自身がその時点で見えているplayer-visible stateを `lisjong-play` の既存board rendererでlive観戦できるようにする。

conceptual flow:

```text
RiichiLab WebSocket ranked
        ↓
lisjong-arena canonical RankedSession / RiichiLabSeatAdapter
        ↓
player-visible decision presentation seam
        ├─> Policy decision / action send   (authoritative execution path)
        └─> lisjong-play live presentation queue
                    ↓
             GuiBoardRenderer
                    ↓
                human viewer
```

Viewerは**完全なread-only observer**であり、Policy判断、transport、合法手、server timingを支配しない。

---

# Critical timing boundary — spectator must never pause ranked execution

Local Spectator #25と最も重要な違い。

RiichiLab rankedにはresponse time budget / deadlineがあるため、GUIのPause / Step / rendering delayでPolicy executionまたはWebSocket responseを止めてはいけない。

```text
Arena ranked execution  ─────────────────────────────> always progresses
                           │      │      │      │
                           v      v      v      v
presentation snapshots    q0     q1     q2     q3
                           ↓
GUI display              follow / pause-display / step-buffer / jump-live
```

したがって:

- Pause = **表示だけをfreeze**。対局は継続する
- Step = pause中に受信済みpresentation frameを1つだけ進める
- Resume / Follow Live = live edgeへ復帰する
- GUI描画遅延やwindow操作をPolicy timingへ逆流させない
- presentation consumerの遅延で`request_action` responseを遅らせない

#25の`SpectatorControl`を意味だけ変えて無理にreuseしない。共通化は実際に同じpresentation semanticsだけに限定する。

---

# Initial launch model

初期scopeでは、**既に別processで走っているArena ranked CLIへ後付けattachしない**。

IPC / local socket / process discoveryを導入せず、lisjong-playの専用entry pointからArenaのsupported ranked runtimeを起動する。

conceptual CLI:

```powershell
lisjong-play-riichilab --profile lisjong-dev --record-dir C:\Dev\lisjong-artifacts\riichilab
```

または同等の `python -m ...` entry point。

exact command/nameはimplementation preflightでcurrent conventionsに合わせる。

このentry pointはArenaのprofile / credential / ranked orchestrationを**再実装しない**。one-game / bounded-run lifecycle、credential resolution、durable record compositionはcurrent Arena contract（#168/#232）をreuseし、lisjong-playはtoken値を保持・表示・persistしない。初期manual smokeはexactly 1 completed hanchanにboundedする。

既存Arena CLIから開始済みの対局へattachする機能はsuccessor scopeとする。

---

# Arena prerequisite — bounded live observer seam

current Arena `RankedSession` は `request_action` をcanonical `RiichiLabSeatAdapter`へ渡し、そのままPolicy判断へ進む。外部presentation consumer向けのsupported observer hookは現在存在しない。

本IssueのためにArena変更が必要な場合は、**bounded prerequisite**として最小のobserver seamを追加する。

要求する性質:

- current canonical RiichiLab parser / adapterを通過したplayer-visible stateだけを公開する
- Policyへ渡すstateと同じdecision prefixにbindする
- presentation consumerがtransport payloadを独自parseしない
- observerはPolicy decision / action mapping / validationを変更しない
- observer callback / queueの遅延がranked response pathをblockしない
- token / Authorization / credential値をobserverへ渡さない
- private `_adapter` / mutable session fieldをconsumerから直接読ませない
- generic project-wide event busを作らない

望ましい最小surfaceは、例えばdecisionごとのimmutable player-visible presentation valueと、必要最小限のlifecycle factである。具体的な型名はpreflightで決める。

```text
start / bound seat
player-visible decision state
our sent action (presentationに必要なら)
action ack status (presentationに必要なら)
end_game / final scores
failure
```

`lisjong-play`側でraw `request_action` JSONやbase64 Observationを独自decodeする方式は採用しない。

---

# Presentation source

RiichiLab live sourceは、Arenaから受けたplayer-visible canonical stateを既存presentationへprojectする。

優先順位:

1. Replay / existing GUIがすでに持つ `PolicyInput` / public board projectionをreuseできるならreuse
2. RiichiLab専用差分だけ薄く追加
3. network viewer専用のMahjong rule/state reconstructionは書かない

表示候補:

- current round / honba / riichi sticks
- scores / seat winds / dealer
- rivers
- melds
- riichi state
- dora indicators
- lisjong自身のconcealed hand
- current decision / request ordinal
- lisjong自身が送ったrecent action
- action ack / connection status
- final scores / rank

RiichiLab protocol / Arena canonical stateがexactに提供しないround-result detail（yaku / han / fu / hidden winning hand等）は推測しない。

---

# Information boundary

本Viewerは **lisjong bot自身のplayer-visible view** だけを表示する。

```text
RiichiLab player-visible observation
        ↓
Arena canonical adapter / decision state
        ├─> Policy
        └─> presentation
```

禁止:

- opponents' concealed handの推測
- server-side / omniscient game logの取得
-複数sourceをmergeしたsynthetic omniscient state
- future draw / wall / dead wall reconstruction
- viewer側のshanten / ukeire / scoring / legality計算
- presentation情報をPolicy inputへfeedbackすること

Local Spectator #25で可能な将来のengine-owned omniscient projectionとは別物である。

---

# Durable record interaction

Arena `--record-dir`との併用を維持する。

```text
same ranked session
    ├─> live presentation (ephemeral, read-only)
    └─> Arena durable ranked record (authoritative persisted source)
```

live viewerのためにdurable record formatを変更しない。

禁止:

- incomplete staging bundleをtailする
- `protocol.jsonl`をlive IPCとして使う
- diagnostic traceをpresentation APIとして扱う
- viewerがrecord bytesへ書き込む

viewer有無で完成したdurable recordのsemantic content / provenanceが変わらないこと。

---

# Threading / failure isolation

- Arena ranked / Policy executionはworker側
- Tk widget操作はmain threadのみ
- observer -> GUI deliveryはbounded / non-blocking
- GUIが遅くてもranked executionを遅らせない
- presentation queueは無制限に増やさない
- cumulative board snapshotは必要に応じてcoalesce可能
- terminal / failure factをsilentに落とさない
- viewer rendering failureをsuccessful match completionとして偽装しない

Window closeは**active ranked gameを暗黙に切断しない**。

初期scopeでは、windowを閉じた場合はpresentationをdetachし、ranked session自体はArenaのauthoritative lifecycleに従って完走させる。途中退出 / resign / intentional disconnect UIは本Issueで追加しない。

---

# Preflight

実装開始時に最低限確認する:

## lisjong-play

- current main / `AGENTS.md`
- README / `docs/architecture.md`
- `GuiBoardRenderer`
- Replay Viewerのplayer-safe board projection
- Local Spectator #25 source / queue / Tk threading
- current presentation/result components

## lisjong-arena

- current main
- `riichilab.session.RankedSession`
- `RiichiLabSeatAdapter`
- ranked runner / transport lifecycle
- profile / credential resolution
- durable ranked record path
- current RiichiEnv / PolicyInput projection

Arenaにsupported observer seamが無いことを再確認してから、必要最小限のcross-repo prerequisiteを実装する。

---

# Tests

## Arena observer boundary

- observer有無で同じfake ranked event streamに対するPolicy responseが同一
- observerへcredential / tokenが渡らない
- exact request / decision prefixへbindされる
- observerの遅いconsumerがPolicy response pathをblockしない設計
- observer failure / detachがranked protocol stateを壊さない
- final scores / terminal failure lifecycleが一意にdeliveryされる

## lisjong-play source

- Arena-provided player-visible state -> existing `GuiBoardView` projection
- opponents' concealed truthを要求しない
- viewer側でlegality / scoring / progressionを再実装しない
- orientationはbound bot seatを基準に安定する
- recent action / status表示がauthoritative factsだけを使う

## Live display control

- Pause display中もranked worker / Policy decisionは進む
- Stepはbuffered presentationをexactly 1 frame進める
- Follow Liveでlatest stateへ復帰する
- slow Tk consumerでもqueue growthがbounded
- window closeでviewerはdetachするがranked gameを暗黙abortしない

## Regression

- Human Play GUI unchanged
- Replay Viewer unchanged
- Local Spectator #25 pause semantics unchanged
- durable ranked record on/off behavior unchanged

network CIは行わず、fake transport / synthetic event streamを正本とする。real RiichiLabはmanual one-hanchan smokeだけに限定する。

---

# Manual smoke

実装・merge後、`lisjong-dev` profileで**1 ranked hanchanだけ**実行する。

確認:

```text
profile = lisjong-dev
Policy = configured current development Policy
viewer renders live player-visible board
no Policy timeout / protocol error attributable to viewer
Pause display does not pause game
Step / Follow Live operate locally
end_game reached
final scores shown
durable record (when enabled) publishes and strict-read succeeds
```

これはstrength evaluationではなく、connectivity / presentation / acquisition integration smokeである。

---

# Non-goals

- already-running external Arena processへのattach / IPC
- RiichiLab server-side spectator APIの実装
- omniscient RiichiLab viewer
- opponents' concealed hands
- live protocol trace tailing
- replay from incomplete durable bundle
- ranked reconnect / retry / automatic requeue
- multiple concurrent games
- web UI
- remote spectator sharing
- AI reasoning / logits / Q-value display
- strength evaluation / rating analysis
- Policy変更
- generic cross-project event bus / Viewer framework
- viewerからのaction override / human takeover

---

# Acceptance criteria

- [ ] current Human / Replay / Local Spectator presentation boundariesをpreflightした
- [ ] Arena ranked runtimeにsupported player-visible observer seamが存在する、またはbounded prerequisiteとして追加した
- [ ] lisjong-playがraw RiichiLab protocol / base64 Observationを独自parseしていない
- [ ] dedicated RiichiLab live spectator entry pointから1 ranked hanchanを開始できる
- [ ] Arena profile / credential resolutionをreuseし、tokenをlisjong-playがpersistしない
- [ ] bound bot seatを基準にpublic board + own concealed handをlive表示できる
- [ ] scores / rivers / melds / riichi / dora / round metadataを既存rendererで表示できる
- [ ] presentation observerがPolicy / action mapping / legality / transport timingを変更しない
- [ ] Pauseはdisplay-onlyで、ranked executionを停止しない
- [ ] paused Step / Follow Liveがbounded buffer上で成立する
- [ ] slow / closed viewerがranked workerをblockまたは暗黙abortしない
- [ ] final scores / terminal stateを表示する
- [ ] unsupported round-result detailを推測しない
- [ ] hidden / future informationがpresentationにもPolicyにも混入しない
- [ ] `--record-dir`併用時もdurable record semantics / strict readbackが不変
- [ ] Human Play / Replay / Local Spectatorの既存semanticsを変更していない
- [ ] fake transport / focused testsを追加し、real network testをCIへ入れていない
- [ ] manual one-hanchan smokeでviewer付きranked完走を確認した

**Comment by lisbun** [comment]
Arena prerequisite is complete.

- lisbun/lisjong-arena#274 merged as `60cf4df92d68ed7bc347bf87d6bd0a2544da7120`
- supported seam is now available via `BoundedRankedPresentationBuffer`
- decision facts expose the exact `PolicyInput` seen by Policy plus canonical selected `InternalAction`
- completion/failure terminal lifecycle is one-of-one and owned by `run_ranked_game()`
- optional presentation is threaded through `acquire_ranked_game_record()` without changing durable record semantics

## lisjong-play implementation direction

Current `lisjong-play` main is `bfed67f`. Its Arena dependency is still pinned to old `e88eacf...`, and its direct lisjong pin (`d39abcd...`) differs from current Arena's `f29d129...`. #41 implementation should therefore update dependency pins coherently, including Arena to merged #274.

Use the existing Replay `PolicyInput -> GuiBoardView` projection as the basis for the live source. Extract only the concrete shared projection needed by Replay and RiichiLab live; do not parse raw RiichiLab `request_action` JSON/base64 in lisjong-play and do not introduce a generic ViewerState/event bus.

For the live controls, do NOT reuse local Spectator's `SpectatorControl` semantics: that object deliberately blocks local engine execution at selector boundaries. RiichiLab live needs presentation-only state:
- Follow Live: show latest retained frame
- Pause: freeze GUI cursor only; Arena worker continues
- Step: advance exactly one already-buffered retained frame while paused
- Follow Live/resume: jump to live edge
- overflow: bounded history, drop/coalesce oldest cumulative decision frames and surface skipped count
- terminal completion/failure must never be silently lost
- window close detaches the presentation buffer/viewer and must not abort the active ranked game

Initial entry point should be dedicated (e.g. `lisjong-play-riichilab`), run exactly one ranked hanchan, reuse Arena `resolve_profile()`, `resolve_credential()`, policy factory, `resolve_ranked_record_path()`, and `acquire_ranked_game_record()` / `run_ranked_game()`. lisjong-play must not persist or display token values.

The bound bot seat should stay at bottom. Display only the bot's own concealed hand plus public board facts from Arena-provided `PolicyInput`; no hidden-hand merge, legality/scoring/progression reconstruction, shanten/ukeire, or unsupported result inference.

After implementation: fake/synthetic tests only in CI; real network smoke remains exactly one `lisjong-dev` hanchan after merge.

**Comment by lisbun** [comment]
PR #42 merged successfully as `5b7920375f0e351f053ac17ec77a3ce7aa2a38cb`.

Issue #41 intentionally remains open. The remaining acceptance step is the merged-main manual smoke: exactly one `lisjong-dev` ranked hanchan using the new RiichiLab live viewer, preferably with `--record-dir`, verifying live board rendering, display-only Pause, Step / Follow Live, end_game / final scores, no viewer-attributable timeout or protocol error, and strict durable record readback.

**Comment by lisbun** [comment]
Merged-main manual smoke completed successfully.

Validation completed:
- PR #42 merged on main as `5b7920375f0e351f053ac17ec77a3ce7aa2a38cb`
- real `lisjong-dev` ranked hanchan completed with the new RiichiLab live viewer
- live spectating succeeded
- durable ranked record was published via `--record-dir`
- strict readback succeeded with `load_ranked_game_record()` / `summarize_ranked_game_record()`
- record identity: `41323cbc7256d3faa9a63c66277c1b74793330b380b52cdd0512aabc3283b4bd`
- readback summary: seat=2, requests=101, responses=101, acknowledged_requests=101, deserialized_observations=101, scores=(-1900, 28500, 28400, 45000)

The post-merge smoke acceptance step is complete. Closing #41 as completed.
